### ISSUE TITLE
Implementing a CommonMark++ parser

### ISSUE THREAD

**Issue by choldgraf** [ExecutableBookProject/meta#18]
This is an issue to discuss whether / how to implement a CommonMark + directives parser for Sphinx, as @chrisjsewell and I had discussed earlier.

## The problem

The `recommonmark` project piggy-backs on the `commonmark-py` project to parse markdown. It then defines a Sphinx parser that sub-classes the `docutils` parser and defines methods that convert the commonmark-py AST into docutils AST (https://github.com/readthedocs/recommonmark/blob/master/recommonmark/parser.py#L21).

Under the hood it's still using docutils methods since they sub-class the docutils parser, and as a result there is some weird behavior (like `nested_parse` expecting rST in the content blocks).

## One solution

@chrisjsewell proposed writing our own CommonMark -> docutils AST parser, and then adding on the syntax for roles and directives. This would be two things:

1. A Sphinx parser that reads in markdown, and uses:
2. Our own Praser that *behaves* like a docutils parser, but under the hood is utilizing a more modern state-machine software (https://github.com/pytransitions/transitions) to parse markdown.

The hope is that this parser would be easier to maintain, understand, and grow as we wished to support new syntax. It would be a collection of "markdown -> docutils AST" rules, rather than relying on an intermediate AST as the `commonmark-py` project does.

## A question - could we continue using `commonmark-py`?

As I was looking through documentation, I am wondering whether we could still use the `commonmark-py` machinery to parse basic commonmark syntax, and then use our own statemachine parser to handle the "extra" grammar elements like roles and directives. 

Basically, I'm wondering whether we could do the same thing that `recommonmark` does, but instead of sub-classing a docutils Parser, we sub-class a parser that knows how to parse *only the subset of markdown that commonmark doesn't cover*.

If this were possible, I feel like we wouldn't need to worry about re-writing the test suite of `commonmark-py`, and we could then focus only on the extra syntax needed for things like roles and directives. We could then also have a markdown parser under the hood for the `nested_parse` sections.

I'm curious what @chrisjsewell thinks about this - mostly I am trying to find ways that we don't have to write our own from-scratch markdown parser as I'm a bit worried about all the edge-cases we'll have to consider :-)

**Comment by choldgraf** [comment]
After thinking through this more, we considered using [mistune](https://github.com/lepture/mistune) to do the rendering bit. Mistune will parse markdown, and then output an AST as a dictionary of parsed elements. It is also extendable, so we could extend the syntax w/ regexes in our own Python classes. We could then do something like:

* Implement a Sphinx parser that uses our own MistuneParser. It would include a few of the methods that the docutils parser uses
* The MistuneParser will do:
   * Parse the text with Mistune + our plugins
   * Maybe take a pass at the Mistune AST to add extra stuff and modify this AST
   * Loop through the AST and convert things into docutils

**Comment by choldgraf** [comment]
I just played around with implementing a plugin for mistune, and it is pretty straightforward, this was enough to get it parsing text between double-quotes:

```python
md = """
# Test one
This is $2+2$ a paragraph.
"""

DOLLAR_INLINE_PATTERN = ("\$(.*?)\$")

# This is what is called if the regex is matched
def parse_dollar_inline(self, m, state):
    text = m.group(1)
    return 'dollar_inline', self.render(text, state)

# This defines how this inline element should make it into the AST
def render_ast_dollar_inline(text):
    text[0]['type'] = 'dollar_inline'
    return text[0]

# This manages the whole registration process
def plugin_dollar_inline(md):
    md.inline.register_rule(
        'dollar_inline', DOLLAR_INLINE_PATTERN, parse_dollar_inline)

    # md.inline.rules is a set of in-build plugins so we just append to it
    # I think it also defines rule precedence
    md.inline.rules.append('dollar_inline')

    # This is just for rendering
    if md.renderer.NAME == 'ast':
        md.renderer.register('dollar_inline', render_ast_dollar_inline)

# Render AST using our custom plugin
renderer = mistune.AstRenderer()
markdown = mistune.create_markdown(renderer=renderer, plugins=[plugin_dollar_block])
ast_mistune = markdown(md)
"""
```

**Comment by chrisjsewell** [comment]
Yeh, eurghh. I looked into retrieving the line/character numbers with mistune (a necessity for an LSP and for error reporting), and I'd say there is no easy way. The structure of the parser really doesn't make it easy.
I looked at mistletoe, and with that, although its not in there now, there is a fighting chance to recover these. Also I noted that they comment on the relative performance of the packages [here](https://github.com/miyuchina/mistletoe#performance):

> We notice that Mistune is the fastest Markdown parser, and by a good margin, which demands some explanation. mistletoe's biggest performance penalty comes from stringently following the CommonMark spec, which outlines a highly context-sensitive grammar for Markdown. Mistune takes a simpler approach to the lexing and parsing process, but this means that it cannot handle more complex cases, e.g., precedence of different types of tokens, escaping rules, etc...

**Comment by choldgraf** [comment]
Interesting - I guess it would make sense that mistune is faster since it basically just says "I ignored a bunch of edgecases so I don't have to check for them".

How much do we care about parsing speed etc? Especially if we're doing a lot of cacheing w/ Sphinx, will this become a big issue? I think mistletoe will be something like 50% the speed of mistune...

Mistletoe was extensible as well, yeah? Also worth noting that I don't believe that mistletoe is maintained much anymore (though the maintainer seemed much friendlier than the mistune person)

**Comment by chrisjsewell** [comment]
To the extent of the difference in speed, I don't think it matters. As we saw in your profiling, a lot more time is taken carrying out directive/role function logic.

Mistletoe is super extensible. Its essentially very much like the Finite-State Machine I wrote, in that each block/inline token is a separate class that deals with its own state. You can add/remove any of the tokens, and control things like the data stored and parsing precedence (see [here](https://github.com/miyuchina/mistletoe#a-new-token)). IMO, its 'theoretically' a lot better than mistune.

But yes the con is that it doesn't appear to be maintained much lately

**Comment by jstac** [comment]
It's concerning when the maintainer doesn't respond to PRs like like [this one](https://github.com/miyuchina/mistletoe/pull/72) or [this one](https://github.com/miyuchina/mistletoe/pull/87).

**Comment by jstac** [comment]
The [marko](https://github.com/frostming/marko) project looks better maintained.  Again, the trade-off between speed and conforming to the exact CommonMark spec is a factor.

**Comment by choldgraf** [comment]
A quick note on execution speed. I tried parsing the [markdown document that @AakashGfude added to the test cases repo](https://github.com/ExecutableBookProject/markup_test_cases/blob/master/wealth_dynamics.md) to see how fast each markup parser took. Here's the quick result:

* **Mistune**: 3.99 ms ± 307 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
* **Mistletoe**: 5.92 ms ± 207 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
* **Commonmark-py**: 7.07 ms ± 556 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
* **Mako**: 26.1 ms ± 2.92 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

And for comparison, here is the result of running a very vanilla docutils parser on the document:

* **docutils**: 108 ms ± 10.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

However, this probably isn't a fair comparison, because it's doing a lot of checking whether directives/roles exist etc.

**Comment by choldgraf** [comment]
I also added this information, plus a short list of pros and cons for each, here:

https://github.com/ExecutableBookProject/meta/wiki/Resources:-Markdown-(MD)#markdown-parsers-in-python

It's in the wiki, so please feel free to update and add information if you think of something

**Comment by choldgraf** [comment]
Another option I hadn't considered, but might be attractive especially if we don't think any of these projects will work for us, is to use a Python wrapper to the `cmark` package. cmark is written in C++ and is super fast and well-maintained:

For one example, here's a lightweight Python wrapper for this package:

https://pypi.org/project/paka.cmark/

It can be used like:

```python
from paka import cmark
import xml.etree.ElementTree as ET

def parse_md_to_xml(md):
    xml = cmark.to_xml(md)
    tree = ET.fromstring(xml)
    return tree

tree = parse_md_to_xml(md)
```

This is about 2-3 times faster than mistune as well...

**Comment by chrisjsewell** [comment]
I'm dead set on mistletoe at this point. Yes, the maintenance is a con, but I think it is just so much more well structured and flexible than the other options.

In https://github.com/chrisjsewell/mistletoe/blob/myst/test.ipynb, you'll see I have forked it, and with just a few code additions, I have implemented the majority of the markdown -> docutils bridge, including capturing block token source code line ranges.

If you are in the office soon we can discuss?
